### ISSUE TITLE
mock_dynamodb2 failed to update dictionary attribute 

### ISSUE THREAD

**Issue by estyxx** [spulec/moto#3577]
Got ` AttributeError: 'NoneType' object has no attribute 'type'` exception while I'm trying to update an attribute on a dynamodb table with moto. 
The attribute is originally created as an empty dictionary after is updated with some values.  

Basically when it is checking if any of the values in the dictionary has changed it is getting `None` from the `original`  dictionary because was empty 

```python
# In moto/dynamodb2/responses.py:828: in _build_updated_new_attributes: 
#  changed    = {'proxy_used': DynamoType: {'M': {'hostname': DynamoType: {'S': 'http://helloworld.com'}}}, 'title': DynamoType: {'S': 'Hello World'}}
# original   = {}

for key in changed.keys()
    if changed[key] != original.get(key, None)
```

but the `DynamoType.__ne__` method it is expecting `other` to be a `DynamoType` instead is getting `None`
```python
    # moto/dynamodb2/models/dynamo_type.py:161 
    # other      = None
    # self       = DynamoType: {'S': 'Hello World'}

    def __ne__(self, other):
>       return self.type != other.type or self.value != other.value

```

Here there is an example to reproduce the issue:
```python
def test_moto():
    import boto3
    from moto import mock_dynamodb2

    with mock_dynamodb2():
        dynamodb = boto3.resource("dynamodb", region_name=AWS_REGION)
        table = dynamodb.create_table(
            TableName="request",
            KeySchema=[
                {"AttributeName": "request_id", "KeyType": "HASH"}
            ],
            AttributeDefinitions=[
                {"AttributeName": "request_id", "AttributeType": "S"},
                {"AttributeName": "url", "AttributeType": "S"},
                {"AttributeName": "retrieved_on_utc", "AttributeType": "S"},
            ],
            GlobalSecondaryIndexes=[
                {
                    "IndexName": "url_date",
                    "KeySchema": [
                        {"AttributeName": "url", "KeyType": "HASH"},  # Partition key
                        {
                            "AttributeName": "retrieved_on_utc",  # Sort key
                            "KeyType": "RANGE",
                        },
                    ],
                    "Projection": {"ProjectionType": "ALL"},
                    "ProvisionedThroughput": {"ReadCapacityUnits": 10, "WriteCapacityUnits": 10},
                }
            ],
            ProvisionedThroughput={"ReadCapacityUnits": 10, "WriteCapacityUnits": 10}
        )
        dynamodb.meta.client.get_waiter("table_exists").wait(TableName="request")
        request_table = dynamodb.Table("request")
        request_table.put_item(
            Item={
                "request_id": "b1d5ebea4cea11eba828acde48001122",
                "url": "https://www.brown.com/",
                "request_parameters": {},
            },
            ReturnValues="NONE",
        )

        # this is working
        request_table.update_item(
            Key={"request_id": "b1d5ebea4cea11eba828acde48001122"},
            UpdateExpression='set meta_data = :val10',
            ExpressionAttributeValues={
                ':val10': {'title': 'Hello World', 'proxy_used': {'hostname': 'http://helloworld.com'}}},
            ReturnValues="UPDATED_NEW",
        )

        # this is raising
        request_table.update_item(
            Key={"request_id": "b1d5ebea4cea11eba828acde48001122"},
            UpdateExpression='set request_parameters = :val1, meta_data = :val10',
            ExpressionAttributeValues={
                ':val1': {'title': 'Hello World', 'proxy_used': {'hostname': 'http://helloworld.com'}},
                ':val10': {'title': 'Hello World', 'proxy_used': {'hostname': 'http://helloworld.com'}}},
            ReturnValues="UPDATED_NEW",
        )

```

Here there is the complete traceback: 
```bash
            request_table.update_item(
                Key={"request_id": "b1d5ebea4cea11eba828acde48001122"},
                UpdateExpression='set request_parameters = :val1, meta_data = :val10',
                ExpressionAttributeValues={
                    ':val1': {'title': 'Hello World', 'proxy_used': {'hostname': 'http://helloworld.com'}},
                    ':val10': {'title': 'Hello World', 'proxy_used': {'hostname': 'http://helloworld.com'}}},
>               ReturnValues="UPDATED_NEW",
            )

../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/boto3/resources/factory.py:520: in do_action
    response = action(self, *args, **kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/boto3/resources/action.py:83: in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/client.py:357: in _api_call
    return self._make_api_call(operation_name, kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/client.py:663: in _make_api_call
    operation_model, request_dict, request_context)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/client.py:682: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/endpoint.py:102: in make_request
    return self._send_request(request_dict, operation_model)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/endpoint.py:137: in _send_request
    success_response, exception):
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/endpoint.py:256: in _needs_retry
    caught_exception=caught_exception, request_dict=request_dict)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/hooks.py:356: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/hooks.py:228: in emit
    return self._emit(event_name, kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/hooks.py:211: in _emit
    response = handler(**kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/retryhandler.py:183: in __call__
    if self._checker(attempts, response, caught_exception):
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/retryhandler.py:251: in __call__
    caught_exception)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/retryhandler.py:269: in _should_retry
    return self._checker(attempt_number, response, caught_exception)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/retryhandler.py:317: in __call__
    caught_exception)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/retryhandler.py:223: in __call__
    attempt_number, caught_exception)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/retryhandler.py:359: in _check_caught_exception
    raise caught_exception
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/endpoint.py:197: in _do_get_response
    responses = self._event_emitter.emit(event_name, request=request)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/hooks.py:356: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/hooks.py:228: in emit
    return self._emit(event_name, kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/botocore/hooks.py:211: in _emit
    response = handler(**kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/core/models.py:323: in __call__
    request, request.url, request.headers
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/core/responses.py:202: in dispatch
    return cls()._dispatch(*args, **kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/core/responses.py:312: in _dispatch
    return self.call_action()
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/core/utils.py:277: in _wrapper
    response = f(*args, **kwargs)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:86: in call_action
    response = getattr(self, endpoint)()
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:814: in update_item
    existing_attributes, item_dict["Attributes"]
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:828: in _build_updated_new_attributes
    for key in changed.keys()
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:829: in <dictcomp>
    if changed[key] != original.get(key, None)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:828: in _build_updated_new_attributes
    for key in changed.keys()
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:829: in <dictcomp>
    if changed[key] != original.get(key, None)
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:828: in _build_updated_new_attributes
    for key in changed.keys()
../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/responses.py:829: in <dictcomp>
    if changed[key] != original.get(key, None)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = DynamoType: {'S': 'Hello World'}, other = None

    def __ne__(self, other):
>       return self.type != other.type or self.value != other.value
E       AttributeError: 'NoneType' object has no attribute 'type'

other      = None
self       = DynamoType: {'S': 'Hello World'}


../../../Library/Caches/pypoetry/virtualenvs/content-engine-T4YngYX3-py3.7/lib/python3.7/site-packages/moto/dynamodb2/models/dynamo_type.py:161: AttributeError

```

Seems something very easy to fix.. think I could make a PR to fix this :)

**Comment by bblommers** [comment]
Hi @estyxx, are you able to upgrade to the latest dev-release (1.3.16.dev184)?
I can reproduce this for 1.3.16, but not for the latest version (master), so I think it was already fixed :)

**Comment by estyxx** [comment]
Hi sorry the late reply, I forgot this... 
I just try with the dev version and the error is gone... so think is fixed! So think I just have to wait for the next release...

**Comment by bblommers** [comment]
Awesome, thanks for letting us know!

**Comment by bblommers** [comment]
@estyxx, moto 2.0.0 is now been released! See #3722
